projects/file-parsing/main.go

```diff
@@ -0,0 +1,360 @@
+	// then loop over the byte slice
+	for startIndex := 2; startIndex < len(file); {
+		// initialise a variable to store the high score
+		var highScore int32
+
+		// parse the high score (signed 32-bit integer) based on the Endianness
+		// (!) but why are we using Uint32 method (when we know its "signed") and then converting it to int32 (presumably "signed").. where is the native xEndian.Int32 method?
+		if bigEndian {
+			highScore = int32(binary.BigEndian.Uint32(file[startIndex : startIndex+4]))
+		} else if littleEndian {
+			highScore = int32(binary.LittleEndian.Uint32(file[startIndex : startIndex+4]))
+		}
+
+		// increment the start pointer by 4
+		startIndex += 4
+
+		// set the end pointer to the start pointer
+		endIndex := startIndex
+
+		// while the value is not a null terminator (deliberately using byte() to be explicit)
+		// walk the end pointer along, to establish the end of the string
+		for file[endIndex] != byte(0x00) {
+			endIndex++
+		}
+
+		// parse the player name (UTF-8)
+		name := string(file[startIndex:endIndex])
+
+		// add the player to the dataSlice
+		*dataSlice = append(*dataSlice, Player{Name: name, HighScore: highScore})
+
+		// adjust the start pointer to skip past the null terminator
+		startIndex = endIndex + 1
+	}
+
+	return nil
+}
+
+func attemptToParse(file []byte) ([]Player, error) {
+	// initialise a slice of Player structs
+	var dataSlice []Player
+
+	// doing this sequentially isn't great...
+	// how to establish what type of file it is before attempting to parse it?
+
+	err := parseJSON(file, &dataSlice)
+	// if err != nil {
+	// 	fmt.Println(err) // temporary
+	// }
+	if err == nil {
+		return dataSlice, nil
+	}
+
+	err = parseRepeatedJSON(file, &dataSlice)
+	// if err != nil {
+	// 	fmt.Println(err) // temporary
+	// }
+	if err == nil {
+		return dataSlice, nil
+	}
+
+	err = parseCSV(file, &dataSlice)
+	// if err != nil {
+	// 	fmt.Println(err) // temporary
+	// }
+	if err == nil {
+		return dataSlice, nil
+	}
+
+	err = parseBinary(file, &dataSlice)
+	// if err != nil {
+	// 	fmt.Println(err) // temporary
+	// }
+	if err == nil {
+		return dataSlice, nil
+	}
+
+	return nil, errors.New("[2] could not parse the file")
+}
+
+func parseFile(filename string) ([]Player, error) {
+	// try to read the file
+	file, err := os.ReadFile(filename)
+
+	// if we can't read the file then error
+	if err != nil {
+		return nil, errors.New("[1] error reading the file")
+	}
+
+	dataSlice, err := attemptToParse(file)
+
+	// if we can't parse the file then error
+	if err != nil {
+		return nil, errors.New("[1] error parsing data from the file")
+	}
+
+	// if the dataSlice is empty then error
+	if len(dataSlice) == 0 {
+		return nil, errors.New("[1] the file likely contains no data")
+	}
+
+	return dataSlice, nil
+}
+
+func getHighestLowestScorePlayers(jsonDataSlice []Player) (Player, Player, error) {
+	// create two variables to hold the highest/lowest scoring players
```

Comment:
They will be nil value integers, eg. `0`
This is a problem if the highest high score is less than 0 and if the lowest low score is larger than 0

Reply:
That's one valid approach. Another would be to use [sentinel values](https://en.wikipedia.org/wiki/Sentinel_value) (i.e. start your lowest at `math.MaxInt` and your highest at `math.MinInt` because then your comparisons will happen to work correctly.